sljit_src/sljitNativeMIPS_common.c

```diff
@@ -390,72 +390,76 @@ static sljit_s32 function_check_is_freg(struct sljit_compiler *compiler, sljit_s
 
-#if !defined(SLJIT_IS_FPU_AVAILABLE) || SLJIT_IS_FPU_AVAILABLE
 static void get_cpu_features(void)
 {
-#if defined(__GNUC__)
+#if !defined(SLJIT_IS_FPU_AVAILABLE) && defined(__GNUC__)
 	sljit_u32 fir = 0;
-#if defined(SLJIT_CONFIG_MIPS_32) && SLJIT_CONFIG_MIPS_32
-#if (!defined(SLJIT_DETECT_FR) || SLJIT_DETECT_FR >= 1) \
-	&& (!defined(SLJIT_MIPS_REV) || (defined(SLJIT_MIPS_REV) && SLJIT_MIPS_REV < 6))
-	sljit_s32 flag = -1;
-#endif
-#endif /* SLJIT_CONFIG_MIPS_32 */
-#endif /* __GNUC__ */
+#endif /* !SLJIT_IS_FPU_AVAILABLE && __GNUC__ */
 	sljit_u32 feature_list = CPU_FEATURE_DETECTED;
 
-#if defined(__GNUC__) || !defined(SLJIT_IS_FPU_AVAILABLE)
+#if defined(SLJIT_IS_FPU_AVAILABLE)
+#if SLJIT_IS_FPU_AVAILABLE
+	feature_list |= CPU_FEATURE_FPU;
+#if SLJIT_IS_FPU_AVAILABLE == 64
+	feature_list |= CPU_FEATURE_FP64;
+#endif /* SLJIT_IS_FPU_AVAILABLE == 64 */
+#endif /* SLJIT_IS_FPU_AVAILABLE */
+#elif defined(__GNUC__)
 	__asm__ ("cfc1 %0, $0" : "=r"(fir));
-	if ((fir & (1 << 22)))
+	if ((fir & (0x3 << 16)) == (0x3 << 16))
 		feature_list |= CPU_FEATURE_FPU;
-#else /* SLJIT_IS_FPU_AVAILABLE */
-	feature_list |= CPU_FEATURE_FPU;
-#endif
 
-	if (!(feature_list & CPU_FEATURE_FPU) && (fir & (1 << 17)) && (fir & (1 << 21)))
-		feature_list |= CPU_FEATURE_FP32;
+#if (defined(SLJIT_CONFIG_MIPS_64) && SLJIT_CONFIG_MIPS_64) \
+	&& (!defined(SLJIT_MIPS_REV) || SLJIT_MIPS_REV < 2)
+	if ((feature_list & CPU_FEATURE_FPU))
+		feature_list |= CPU_FEATURE_FP64;
+#else /* SLJIT_CONFIG_MIPS32 || SLJIT_MIPS_REV >= 2 */
+	if ((fir & (1 << 22)))
+		feature_list |= CPU_FEATURE_FP64;
+#endif /* SLJIT_CONFIG_MIPS_64 && SLJIT_MIPS_REV < 2 */
+#endif /* SLJIT_IS_FPU_AVAILABLE */
 
+	if ((feature_list & CPU_FEATURE_FPU) && (feature_list & CPU_FEATURE_FP64)) {
 #if defined(SLJIT_CONFIG_MIPS_32) && SLJIT_CONFIG_MIPS_32
-	if ((feature_list & CPU_FEATURE_FPU) && (!(feature_list & CPU_FEATURE_FP32))) {
 #if defined(SLJIT_MIPS_REV) && SLJIT_MIPS_REV >= 6
 		feature_list |= CPU_FEATURE_FR;
 #elif defined(SLJIT_DETECT_FR) && SLJIT_DETECT_FR == 0
 #if defined(SLJIT_MIPS_REV) && SLJIT_MIPS_REV >= 5
 		feature_list |= CPU_FEATURE_FR;
 #endif /* SLJIT_MIPS_REV >= 5 */
 #else
-#ifdef PR_GET_FP_MODE
+		sljit_s32 flag = -1;
+#ifndef FR_GET_FP_MODE
+		sljit_f64 zero = 0.0;
+#else /* PR_GET_FP_MODE */
 		flag = prctl(PR_GET_FP_MODE);
+
 		if (flag > 0)
 			feature_list |= CPU_FEATURE_FR;
 #endif /* FP_GET_PR_MODE */
 #if ((defined(SLJIT_DETECT_FR) && SLJIT_DETECT_FR == 2) \
 	|| (!defined(PR_GET_FP_MODE) && (!defined(SLJIT_DETECT_FR) || SLJIT_DETECT_FR >= 1))) \
 	&& (defined(__GNUC__) && (defined(__mips) && __mips >= 2))
 		if (flag < 0) {
-			SLJIT_ASSERT(freg_map[TMP_FREG3] == 16);
-			__asm__ (
-				".data\n"
-				"0:\n"
-				"	.quad 0\n"
-				".text\n"
-				".set oddspreg\n"
-				"	mtc1 %0, $f17\n"
-				"	ldc1 $f16, 0b\n"
-				"	nop\n"
-				"	mfc1 %0, $f17\n"
-			: "+r" (flag) : : "$f16", "$f17");
+			/* TMP_FREG3 is expected to be mapped to register f16 */
```

Comment:
The point is, it does not matter which fpu register you use. The comment makes it look like it does. I would use the typical function return value (`f0` `f1` I think).

Reply:
Fair enough, but then I will rather remove the comment than use a register pair that "might" need saving/restoring.  f16/f17 is a "temporary register" per the n32 ABI which is indirectly why the comment meant.

Reply:
Ok, then mention that f16 is an assembler temporary in the MIPS ABI. Anyway since "$f16", "$f17" is passed as clobbered, the compiler will not use them for anything. The function itself does not use floating point at all.